aten/src/ATen/native/EmbeddingBag.cpp

```diff
@@ -461,6 +461,77 @@ Tensor _embedding_bag_dense_backward_cpu(const Tensor &grad_, const Tensor &indi
   return index_grad_weight;
 }
 
+template<typename scalar_t>
+Tensor _embedding_bag_per_sample_weights_backward_cpu_template(
+    const Tensor& grad,
+    const Tensor& weight,  // NB: embedding table, not per_sample_weights
+    const Tensor& indices,
+    const Tensor& offset2bag,
+    int64_t mode) {
+  AT_CHECK(
+      mode == MODE_SUM,
+      "embedding_bag_backward: per_sample_weights only supported for mode='sum'");
+
+  AT_ASSERT(grad.dim() == 2)
+  auto embedding_features = grad.size(1);
+
+  AT_ASSERT(indices.dim() == 1);
+  auto num_samples = indices.size(0);
+
+  AT_ASSERT(weight.dim() == 2);
+  AT_ASSERT(weight.size(1) == embedding_features);
+
+  auto output = at::zeros({num_samples}, grad.options());
+
+  auto indices_arg = TensorArg(indices, "indices", 1);
+  checkScalarType("embedding_bag", indices_arg, kLong);
+  checkContiguous("embedding_bag", indices_arg);
+  auto offset2bag_arg = TensorArg(offset2bag, "offset2bag", 1);
+  checkScalarType("embedding_bag", offset2bag_arg, kLong);
+  checkContiguous("embedding_bag", offset2bag_arg);
+
+  auto grad_data = grad.data<scalar_t>();
+  auto grad_stride0 = grad.stride(0);
+  auto grad_stride1 = grad.stride(1);
+
+  auto weight_data = weight.data<scalar_t>();
+  auto weight_stride0 = weight.stride(0);
+  auto weight_stride1 = weight.stride(1);
+
+  auto indices_data = indices.data<int64_t>();
+
+  // The following are contiguous
+  auto output_data = output.data<scalar_t>();
+  auto offset2bag_data = offset2bag.data<int64_t>();
+
+  // XXX: 64 was arbitrarily chosen. There is probably a sweet spot for this number.
+  #pragma omp parallel for if (num_samples >= 64)
+  for (int64_t sample_idx = 0; sample_idx < num_samples; sample_idx++) {
+    auto bag_idx = offset2bag_data[sample_idx];
+    auto embedding_idx = indices_data[sample_idx];
+
+    output_data[sample_idx] = THBlas_dot<scalar_t>(
```

Comment:
The reason this was used initially was due to overhead of the corresponding aten function. It might be wroth reevaluating whether this is still necessary.

Reply:
I think calling BLAS's dot function should be faster than me writing a for loop to manually do the dot product. I can't use PyTorch's dot operator for a few other reasons, though.